source/common/upstream/health_checker_base_impl.cc

```diff
@@ -179,16 +186,19 @@ HealthCheckerImplBase::ActiveHealthCheckSession::~ActiveHealthCheckSession() {
 void HealthCheckerImplBase::ActiveHealthCheckSession::handleSuccess() {
   // If we are healthy, reset the # of unhealthy to zero.
   num_unhealthy_ = 0;
+  ++num_healthy_;
 
   HealthTransition changed_state = HealthTransition::Unchanged;
   if (host_->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC)) {
     // If this is the first time we ever got a check result on this host, we immediately move
     // it to healthy. This makes startup faster with a small reduction in overall reliability
     // depending on the HC settings.
-    if (first_check_ || ++num_healthy_ == parent_.healthy_threshold_) {
+    if (first_check_ || num_healthy_ == parent_.healthy_threshold_) {
       host_->healthFlagClear(Host::HealthFlag::FAILED_ACTIVE_HC);
       parent_.incHealthy();
       changed_state = HealthTransition::Changed;
+    } else {
+      changed_state = HealthTransition::ChangePending;
```

Comment:
I specifically wanted to respect the former semantics of `changed_state`, when it was a boolean.

`Changed` is only set where before `changed_state` would be set to true. Given that all places that used to check for truth-ness of `changed_state` have been changed to `changed_state == HealthTransition::Changed` in a previous PR, truth-ness checks still hold:
```
envoy-dev@envoy-dev:~/src/envoy$ egrep -R '(== HealthTransition::Changed)' . 2>/dev/null
./source/common/upstream/upstream_impl.cc:        if (changed_state == HealthTransition::Changed) {
./source/common/upstream/health_checker_base_impl.cc:      base_time_ms = (changed_state == HealthTransition::Changed ||
./source/common/upstream/health_checker_base_impl.cc:      base_time_ms = (changed_state == HealthTransition::Changed ||
./source/common/upstream/cluster_manager_impl.cc:          if (changed_state == HealthTransition::Changed &&
```

Where `changed_state` would have been false, it's now either `ChangePending` or `Unchanged`. Therefore new state would only affect false-ness checks. Since there are no such checks, the addition of the new state shouldn't be a problem:
```
envoy-dev@envoy-dev:~/src/envoy$ egrep -R '(!= HealthTransition::Changed|== HealthTransition::Unchanged)' . 2>/dev/null
envoy-dev@envoy-dev:~/src/envoy$
```

With the above in mind, the semantics should still hold.